### PR TITLE
Filterable Search Enhancements

### PR DESCRIPTION
**Pull request by alexm118**
Enhance filterable search by expanding the ngram tokenizer we use for matching terms, as well as introducing text based searching against the content, preview description, and topics of the results. 

---

## Additions

- New tokenizer that isn't shared with autocomplete - expands the size of which we can match substrings on search results.
- New searchable fields - content, preview description, and tag names.

## Changes

- Now uses a multimatch search to query across several fields instead of just the title field.

## How to test this PR

1. Reindex your elasticsearch cluster
2. Search a term not found in the title but found in the description (I found blogs on pandemic end up resulting in more results)
3. Confirm you see additional results based on new matching logic.

## Notes and todos

-

## Checklist

- [ ] PR has an informative and human-readable title
  - PR titles are used to generate the change log in [releases](../../releases); good ones make that easier to scan.
  - Consider prefixing, e.g., "Mega Menu: fix layout bug", or "Docs: Update Docker installation instructions".
- [ ] Changes are limited to a single goal (no scope creep)
- [ ] Code follows the standards laid out in the [CFPB development guidelines](https://github.com/cfpb/development)
- [ ] Future todos are captured in comments and/or tickets
- [ ] Project documentation has been updated, potentially one or more of:
  - [This repo’s docs](https://cfpb.github.io/consumerfinance.gov/) (edit the files in the `/docs` folder) – for basic, close-to-the-code docs on working with this repo
  - CFGOV/platform wiki on GHE – for internal CFPB developer guidance
  - CFPB/hubcap wiki on GHE – for internal CFPB design and content guidance

### CODE REVIEW

File: cfgov/v1/documents.py [hunk 1]
```diff
@@ -12,11 +18,25 @@
...
+    'filterable_ngram_tokenizer',
+    tokenizer=tokenizer(
+        'filterable_tokenizer',
+        'ngram',
+        min_gram=1,
+        max_gram=16,
+        token_chars=["letter", "digit", "symbol"]
+    ),
+    filter=['lowercase', token_filter('ascii_fold', 'asciifolding')]
+)
```
Comment (willbarton) [inline cfgov/v1/documents.py:31]: Could or should this live in `search.elasticsearch_helpers` along side the existing `ngram_tokenizer` we have there, with naming/commentary on what's different between them?

**Review by willbarton — APPROVED** [review]
I think this looks good on a read-through.

I have one minor question on code organization/naming.

File: cfgov/v1/documents.py (hunk 1 above)
Comment (alexm118) [reply cfgov/v1/documents.py:31]: I debated that - I initially left it here as I didn't have a good plan for explaining the difference between the two so I felt grouping this analyzer with its usage made more sense. I guess the main difference would be the size of the tokens, as well as the addition of symbols to this analyzer.

File: cfgov/v1/documents.py [hunk 2]
```diff
@@ -116,9 +154,16 @@ def filter_archived(self, search):
         return search.filter("terms", is_archived=self.archived)
 
     def search_title(self, search):
-        return search.query(
-            "match", title={"query": self.title, "operator": "AND"}
-        )
+        if flag_enabled('EXPAND_FILTERABLE_LIST_SEARCH'):
+            query = MultiMatch(
+                query=self.title,
```
Comment (higs4281) [inline cfgov/v1/documents.py:159]: I'm not getting expected results from the title search with the flag enabled.

Example: I would expect, with the 'EXPAND_FILTERABLE_LIST_SEARCH' flag enabled, that content from the `preview_description` field would be indexed, but a search for "pagkalat," which is in the search description field of live Blog page 13848, doesn't show up in search. 

I wonder if this line should be 


```suggestion
                query=search,
```

**Review by higs4281 — COMMENTED** [review]
One question about the title search.

File: cfgov/v1/documents.py (hunk 2 above)
Comment (higs4281) [reply cfgov/v1/documents.py:159]: Ah, I see self.title is the search title param. Still, I'm not getting why the search description term isn't found.

Also, is it intended behavior that the "item name" search only matches an exact phrase in a title?

File: cfgov/v1/documents.py [hunk 3]
```diff
@@ -34,6 +54,8 @@ class FilterablePagesDocument(Document):
     statuses = fields.KeywordField()
     initial_filing_date = fields.DateField()
     model_class = fields.KeywordField()
+    content = fields.TextField()
+    preview_description = fields.TextField()
```
Comment (chosak) [inline cfgov/v1/documents.py:58]: Is it deliberate that only these fields are included in the search? Some of the filterable pages include other fields; for example, DocumentDetailPage, EventPage, and LearnPage include a [`header`](https://github.com/cfpb/consumerfinance.gov/blob/12bcc7de85ab15ca9c9961ff7724990e745c3680/cfgov/v1/models/learn_page.py#L30-L34) field, EventPage includes several other [body fields](https://github.com/cfpb/consumerfinance.gov/blob/12bcc7de85ab15ca9c9961ff7724990e745c3680/cfgov/v1/models/learn_page.py#L178-L182), etc.

File: cfgov/v1/documents.py (hunk 2 above)
Comment (higs4281) [reply cfgov/v1/documents.py:159]: Do I have this right? SublandingFilterablePage (/about-us/blog/) doesn't hit Elasticsearch? The FilterableListMixin just gets querysets?

Comment (alexm118) [reply cfgov/v1/documents.py:159]: So you will need two flags on - `EXPAND_FILTERABLE_LIST_SEARCH` and `ELASTICSEARCH_FILTERABLE_LISTS` both enabled. 

We also don't explicitly index the search_description but rather the preview_description (I think they are identical values in most cases though)

File: cfgov/v1/documents.py (hunk 3 above)
Comment (alexm118) [reply cfgov/v1/documents.py:58]: These we're the fields specified by Sonna as the initial ask

File: cfgov/v1/documents.py (hunk 2 above)
Comment (higs4281) [reply cfgov/v1/documents.py:159]: All questions cleared up after enabling the 'ELASTICSEARCH_FILTERABLE_LISTS' flag as well as 'EXPAND_FILTERABLE_LIST_SEARCH'.

I'm now seeing expected results from the search_description field, including my test case, "pagkalat," and multi-term searches.

**Review by higs4281 — APPROVED** [review]

File: cfgov/v1/documents.py (hunk 1 above)
Comment (alexm118) [reply cfgov/v1/documents.py]: Ended up not using the analyzer so deleting it - using a match_phrase_prefix query now.

File: cfgov/cfgov/settings/base.py [hunk 4]
```diff
@@ -767,7 +767,9 @@
     # Hide archive filter options in the filterable UI
     "HIDE_ARCHIVE_FILTER_OPTIONS": [],
     # Enable ES as the backend for FilterableLists
-    "ELASTICSEARCH_FILTERABLE_LISTS": []
+    "ELASTICSEARCH_FILTERABLE_LISTS": [],
+    # Expand ES Filterable List Search
+    "EXPAND_FILTERABLE_LIST_SEARCH": []
```
Comment (willbarton) [inline cfgov/cfgov/settings/base.py:772]: Incredibly minor thing, can we add a trailing `,` here?
```suggestion
    "EXPAND_FILTERABLE_LIST_SEARCH": [],
```

**Review by willbarton — APPROVED** [review]
One minor comment, but I think this looks good to me on a read-through.

File: cfgov/v1/documents.py [hunk 5]
```diff
@@ -135,6 +170,7 @@ def order_results(self, search):
         if not self.title:
             return search.sort('-date_published')[0:total_results]
         else:
+            print(search.to_dict())
```
Comment (Scotchester) [inline cfgov/v1/documents.py:173]: Debugging statement left in? It's failing the unit tests for writing to stdout.

Comment (Scotchester) [reply cfgov/v1/documents.py:173]: Other than that, was this waiting on anything else?

**Pull request merged**